tools/indexer/wallet/src/main.rs

```diff
@@ -0,0 +1,221 @@
+    dotenv().ok();
+
+    let database_url = env::var("DATABASE_URL").expect("DATABASE_URL must be set");
+    let manager = ConnectionManager::<PgConnection>::new(&database_url);
+    Pool::builder().build(manager).expect(&format!("Error connecting to {}", database_url))
+}
+
+async fn handle_genesis_public_keys(near_config: near_indexer::NearConfig) {
+    let pool = establish_connection();
+    let genesis_height = near_config.genesis.config.genesis_height.clone();
+    let access_keys = near_config.genesis.records.as_ref()
+        .iter()
+        .filter(|record| match record {
+        near_indexer::near_primitives::state_record::StateRecord::AccessKey { .. } => true,
+        _ => false,
+    })
+        .filter_map(|record| if let near_indexer::near_primitives::state_record::StateRecord::AccessKey { account_id, public_key, access_key} = record {
+        Some(AccessKey {
+            public_key: public_key.to_string(),
+            account_id: account_id.to_string(),
+            action: ADD,
+            status: SUCCESS,
+            receipt_hash: "genesis".to_string(),
+            block_height: BigDecimal::from_u64(genesis_height).expect("genesis_height must be provided"),
+            permission: match access_key.permission {
+                near_indexer::near_primitives::account::AccessKeyPermission::FullAccess => FULL_ACCESS,
+                near_indexer::near_primitives::account::AccessKeyPermission::FunctionCall(_) => FUNCTION_CALL,
+            }
+        })
+    } else { None }).collect::<Vec<AccessKey>>();
+
+    diesel::insert_into(schema::access_keys::table)
+        .values(access_keys)
+        .on_conflict_do_nothing()
+        .execute_async(&pool)
+        .await
+        .unwrap();
+}
+
+async fn listen_blocks(mut stream: mpsc::Receiver<near_indexer::BlockResponse>) {
+    let pool = establish_connection();
+
+    while let Some(block) = stream.recv().await {
+        eprintln!("Block height {:?}", block.block.header.height);
+
+        // Handle outcomes
+        let receipt_outcomes = &block
+            .outcomes
+            .iter()
+            .filter_map(|outcome| match outcome {
+                near_indexer::Outcome::Receipt(execution_outcome) => Some(execution_outcome),
+                _ => None,
+            })
+            .collect::<Vec<&near_indexer::near_primitives::views::ExecutionOutcomeWithIdView>>();
+
+        diesel::update(
+            schema::access_keys::table.filter(
+                schema::access_keys::dsl::receipt_hash.eq_any(
+                    receipt_outcomes
+                        .iter()
+                        .filter(|outcome| match &outcome.outcome.status {
+                            near_indexer::near_primitives::views::ExecutionStatusView::Failure(
+                                _,
+                            ) => true,
+                            _ => false,
+                        })
+                        .map(|outcome| outcome.id.clone().to_string())
```

Comment:
1. `.filter_map` will do the job nicely here.
2. Why do you `.clone()` + `.to_string()`? Seems redundant to me.